File: common/test/java/org/openqa/selenium/CookieImplementationTest.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package org.openqa.selenium;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.openqa.selenium.Ignore.Driver.SELENESE;

import org.openqa.selenium.environment.GlobalTestEnvironment;
import org.openqa.selenium.environment.webserver.AppServer;
import static org.openqa.selenium.Ignore.Driver.IE;
import static org.openqa.selenium.Ignore.Driver.REMOTE;

import java.net.URI;
import java.util.Date;
import java.util.Iterator;
import java.util.Random;
import java.util.Set;

public class CookieImplementationTest extends AbstractDriverTestCase {

  @Override
  protected void setUp() throws Exception {
    super.setUp();

    driver.get(pages.simpleTestPage);
    driver.manage().deleteAllCookies();
  }

  @JavascriptEnabled
  @Ignore(SELENESE)
  public void testShouldGetCookieByName() {
    String key = String.format("key_%d", new Random().nextInt());
    ((JavascriptExecutor) driver).executeScript("document.cookie = arguments[0] + '=set';", key);

    Cookie cookie = driver.manage().getCookieNamed(key);
    assertEquals("set", cookie.getValue());
  }

  @JavascriptEnabled
  @Ignore(SELENESE)
  public void testShouldBeAbleToAddCookie() {
    String key = String.format("key_%d", new Random().nextInt());
    Cookie cookie = new Cookie(key, "foo");

    ((JavascriptExecutor) driver).executeScript("return document.cookie");

    driver.manage().addCookie(cookie);

    String current = (String) ((JavascriptExecutor) driver).executeScript("return document.cookie");
    assertTrue(current.contains(key));
  }

  @Ignore(SELENESE)
  public void testGetAllCookies() {
    Random random = new Random();
    String key1 = String.format("key_%d", random.nextInt());
    String key2 = String.format("key_%d", random.nextInt());

    Set<Cookie> cookies = driver.manage().getCookies();
    int count = cookies.size();

    Cookie one = new Cookie(key1, "value");
    Cookie two = new Cookie(key2, "value");

    driver.manage().addCookie(one);
    driver.manage().addCookie(two);

    driver.get(pages.simpleTestPage);
    cookies = driver.manage().getCookies();
    assertEquals(count + 2, cookies.size());

    assertTrue(cookies.contains(one));
    assertTrue(cookies.contains(two));
  }

  @JavascriptEnabled
  @Ignore(SELENESE)
  public void testDeleteAllCookies() {
    ((JavascriptExecutor) driver).executeScript("document.cookie = 'foo=set';");
    int count = driver.manage().getCookies().size();
    assertTrue(count > 0);

    driver.manage().deleteAllCookies();

    Set<Cookie> cookies = driver.manage().getCookies();
    count = cookies.size();
    assertTrue(cookies.toString(), count == 0);
  }

  @JavascriptEnabled
  @Ignore(SELENESE)
  public void testDeleteCookieWithName() {
    Random random = new Random();
    String key1 = String.format("key_%d", random.nextInt());
    String key2 = String.format("key_%d", random.nextInt());

    ((JavascriptExecutor) driver).executeScript("document.cookie = arguments[0] + '=set';", key1);
    ((JavascriptExecutor) driver).executeScript("document.cookie = arguments[0] + '=set';", key2);

    assertNotNull(driver.manage().getCookieNamed(key1));
    assertNotNull(driver.manage().getCookieNamed(key2));

    driver.manage().deleteCookieNamed(key1);

    Set<Cookie> cookies = driver.manage().getCookies();

    assertNull(cookies.toString(), driver.manage().getCookieNamed(key1));
    assertNotNull(cookies.toString(), driver.manage().getCookieNamed(key2));
  }

  @Ignore(SELENESE)
  public void testShouldNotDeleteCookiesWithASimilarName() {
    String cookieOneName = "fish";
    Cookie cookie1 = new Cookie(cookieOneName, "cod");
    Cookie cookie2 = new Cookie(cookieOneName + "x", "earth");
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);
    options.addCookie(cookie2);

    options.deleteCookieNamed(cookieOneName);
    Set<Cookie> cookies = options.getCookies();

    assertFalse(cookies.toString(), cookies.contains(cookie1));
    assertTrue(cookies.toString(), cookies.contains(cookie2));
  }

  @Ignore(SELENESE)
  public void testAddCookiesWithDifferentPathsThatAreRelatedToOurs() {
    driver.get(pages.simpleTestPage);
    driver.manage().deleteAllCookies();

    Cookie cookie1 = new Cookie("fish", "cod", "/common/animals");
    Cookie cookie2 = new Cookie("planet", "earth", "/common/");
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);
    options.addCookie(cookie2);

    AppServer appServer = GlobalTestEnvironment.get().getAppServer();
    driver.get(appServer.whereIs("animals"));

    Set<Cookie> cookies = options.getCookies();
    assertTrue(cookies.contains(cookie1));
    assertTrue(cookies.contains(cookie2));

    driver.get(appServer.whereIs(""));
    cookies = options.getCookies();
    assertFalse(cookies.toString(), cookies.contains(cookie1));
    assertTrue(cookies.toString(), cookies.contains(cookie2));
  }

  @Ignore(SELENESE)
  public void testCanSetCookiesOnADifferentPathOfTheSameHost() {
    Cookie cookie1 = new Cookie("fish", "cod", "/common/animals");
    Cookie cookie2 = new Cookie("planet", "earth", "/common/galaxy");

    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);
    options.addCookie(cookie2);

    AppServer appServer = GlobalTestEnvironment.get().getAppServer();
    driver.get(appServer.whereIs("animals"));
    Set<Cookie> cookies = options.getCookies();

    assertTrue(cookies.toString(), cookies.contains(cookie1));
    assertFalse(cookies.toString(), cookies.contains(cookie2));

    driver.get(appServer.whereIs("galaxy"));
    cookies = options.getCookies();
    assertFalse(cookies.contains(cookie1));
    assertTrue(cookies.contains(cookie2));
  }

  @Ignore(SELENESE)
  public void testShouldNotBeAbleToSetDomainToSomethingThatIsUnrelatedToTheCurrentDomain() {
    Cookie cookie1 = new Cookie("fish", "cod");
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);

    String url = GlobalTestEnvironment.get().getAppServer().whereElseIs("simpleTest.html");
    driver.get(url);

    Cookie cookie = options.getCookieNamed("fish");
    assertNull(String.valueOf(cookie), cookie);
  }

  @Ignore({SELENESE, IE})
  public void testShouldBeAbleToAddToADomainWhichIsRelatedToTheCurrentDomain() {
    String name = gotoValidDomainAndClearCookies();
    if (name == null || name.matches("\\d{1,3}(?:\\.\\d{1,3}){3}")) {
      System.out.println("Skipping test: unable to find domain name to use");
      return;
    }

    assertNull(driver.manage().getCookieNamed("name"));

    String shorter = name.replaceFirst(".*?\\.", "");
    Cookie cookie =
        new Cookie("name", "value", shorter, "/", new Date(System.currentTimeMillis() + 100000));

    driver.manage().addCookie(cookie);

    assertNotNull(driver.manage().getCookieNamed("name"));
  }

  @Ignore({REMOTE, SELENESE, IE})
  public void testShouldBeAbleToIncludeLeadingPeriodInDomainName() throws Exception {
    String name = gotoValidDomainAndClearCookies();
    if (name == null || name.matches("\\d{1,3}(?:\\.\\d{1,3}){3}")) {
      System.out.println("Skipping test: unable to find domain name to use");
      return;
    }
    driver.manage().deleteAllCookies();

    assertNull("Looks like delete all cookies doesn't", driver.manage().getCookieNamed("name"));

    // Replace the first part of the name with a period
    String shorter = name.replaceFirst(".*?\\.", ".");
    Cookie cookie =
        new Cookie("name", "value", shorter, "/", new Date(System.currentTimeMillis() + 100000));

    driver.manage().addCookie(cookie);

    assertNotNull(driver.manage().getCookieNamed("name"));
  }

  @Ignore(SELENESE)
  public void testGetCookieDoesNotRetriveBeyondCurrentDomain() {
    Cookie cookie1 = new Cookie("fish", "cod");
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);

    String url = GlobalTestEnvironment.get().getAppServer().whereElseIs("");
    driver.get(url);

    Set<Cookie> cookies = options.getCookies();
    assertFalse(cookies.contains(cookie1));
  }

  @Ignore({IE, SELENESE})
  public void testShouldBeAbleToSetDomainToTheCurrentDomain() throws Exception {
    URI url = new URI(driver.getCurrentUrl());
    String host = url.getHost() + ":" + url.getPort();

    Cookie cookie1 = new Cookie.Builder("fish", "cod").domain(host).build();
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);

    driver.get(pages.javascriptPage);
    Set<Cookie> cookies = options.getCookies();
    assertTrue(cookies.contains(cookie1));
  }

  @Ignore(SELENESE)
  public void testShouldWalkThePathToDeleteACookie() {
    Cookie cookie1 = new Cookie("fish", "cod");
    driver.manage().addCookie(cookie1);

    driver.get(pages.childPage);
    Cookie cookie2 = new Cookie("rodent", "hamster", "/common/child");
    driver.manage().addCookie(cookie2);

    driver.get(pages.grandchildPage);
    Cookie cookie3 = new Cookie("dog", "dalmation", "/common/child/grandchild/");
    driver.manage().addCookie(cookie3);

    driver.get(GlobalTestEnvironment.get().getAppServer().whereIs("child/grandchild"));
    driver.manage().deleteCookieNamed("rodent");

    assertNull(driver.manage().getCookies().toString(), driver.manage().getCookieNamed("rodent"));

    Set<Cookie> cookies = driver.manage().getCookies();
    assertEquals(2, cookies.size());
    assertTrue(cookies.contains(cookie1));
    assertTrue(cookies.contains(cookie3));

    driver.manage().deleteAllCookies();
    driver.get(pages.grandchildPage);

    cookies = driver.manage().getCookies();
    assertEquals(0, cookies.size());
  }

  @Ignore({IE, SELENESE})
  public void testShouldIgnoreThePortNumberOfTheHostWhenSettingTheCookie() throws Exception {
    URI uri = new URI(driver.getCurrentUrl());
    String host = String.format("%s:%d", uri.getHost(), uri.getPort());

    assertNull(driver.manage().getCookieNamed("name"));
    Cookie cookie = new Cookie.Builder("name", "value").domain(host).build();
    driver.manage().addCookie(cookie);

    assertNotNull(driver.manage().getCookieNamed("name"));
  }

  @Ignore({SELENESE, IE})
  public void testCookieIntegrity() {
    String url = GlobalTestEnvironment.get().getAppServer().whereElseIs("animals");

    driver.get(url);
    driver.manage().deleteAllCookies();

    long time = System.currentTimeMillis() + (60 * 60 * 24);
    Cookie cookie1 = new Cookie("fish", "cod", "/common/animals", new Date(time));
    WebDriver.Options options = driver.manage();
    options.addCookie(cookie1);

    Set<Cookie> cookies = options.getCookies();
    Iterator<Cookie> iter = cookies.iterator();
    Cookie retrievedCookie = null;
    while (iter.hasNext()) {
      Cookie temp = iter.next();

      if (cookie1.equals(temp)) {
        retrievedCookie = temp;
        break;
      }
    }

    assertNotNull(retrievedCookie);
    //Cookie.equals only compares name, domain and path
    assertEquals(cookie1, retrievedCookie);
  }

  @Ignore(SELENESE)
  private String gotoValidDomainAndClearCookies() {
    AppServer appServer = GlobalTestEnvironment.get().getAppServer();

    String name = null;
    String hostname = appServer.getHostName();
    if (hostname.matches("\\w+\\.\\w+.*")) {
      name = hostname;
      driver.get(appServer.whereIs("simpleTest.html"));
    }
    hostname = appServer.getAlternateHostName();
    if (name == null && hostname.matches("\\w+\\.\\w+.*")) {
      name = hostname;
      driver.get(appServer.whereElseIs("simpleTest.html"));
    }

    driver.manage().deleteAllCookies();

    return name;
  }
}

File: dotnet/test/WebDriver.Common.Tests/CookieImplementationTest.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;
using NUnit.Framework;
using OpenQA.Selenium.Environment;

namespace OpenQA.Selenium
{
    [TestFixture]
    public class CookieImplementationTest : DriverTestFixture
    {
        [SetUp]
        public void GoToSimplePageAndDeleteCookies()
        {
            driver.Url = simpleTestPage;
            driver.Manage().Cookies.DeleteAllCookies();
        }

        [Test]
        [Category("JavaScript")]
        public void ShouldGetCookieByName()
        {
            string key = string.Format("key_{0}", new Random().Next());
            ((IJavaScriptExecutor)driver).ExecuteScript("document.cookie = arguments[0] + '=set';", key);

            Cookie cookie = driver.Manage().Cookies.GetCookieNamed(key);
            Assert.AreEqual("set", cookie.Value);
        }

        [Test]
        [Category("JavaScript")]
        public void ShouldBeAbleToAddCookie()
        {
            string key = string.Format("key_{0}", new Random().Next());
            Cookie cookie = new Cookie(key, "foo");

            ((IJavaScriptExecutor)driver).ExecuteScript("return document.cookie");

            driver.Manage().Cookies.AddCookie(cookie);

            string current = (string)((IJavaScriptExecutor)driver).ExecuteScript("return document.cookie");
            Assert.IsTrue(current.Contains(key));
            Assert.That(driver.Manage().Cookies.AllCookies.Contains(cookie), "Cookie was not added successfully");
        }

        [Test]
        public void GetAllCookies()
        {
            Random random = new Random();
            string key1 = string.Format("key_{0}", random.Next());
            string key2 = string.Format("key_{0}", random.Next());

            ReadOnlyCollection<Cookie> cookies = driver.Manage().Cookies.AllCookies;
            int count = cookies.Count;

            Cookie one = new Cookie(key1, "value");
            Cookie two = new Cookie(key2, "value");

            driver.Manage().Cookies.AddCookie(one);
            driver.Manage().Cookies.AddCookie(two);

            driver.Url = simpleTestPage;
            cookies = driver.Manage().Cookies.AllCookies;
            Assert.AreEqual(count + 2, cookies.Count);

            Assert.IsTrue(cookies.Contains(one));
            Assert.IsTrue(cookies.Contains(two));
        }

        [Test]
        [Category("JavaScript")]
        public void DeleteAllCookies()
        {
            ((IJavaScriptExecutor)driver).ExecuteScript("document.cookie = 'foo=set';");
            int count = driver.Manage().Cookies.AllCookies.Count;
            Assert.Greater(count, 0);

            driver.Manage().Cookies.DeleteAllCookies();

            count = driver.Manage().Cookies.AllCookies.Count;
            Assert.AreEqual(0, count);
        }

        [Test]
        [Category("JavaScript")]
        public void DeleteCookieWithName()
        {
            Random random = new Random();
            string key1 = string.Format("key_{0}", random.Next());
            string key2 = string.Format("key_{0}", random.Next());

            ((IJavaScriptExecutor)driver).ExecuteScript("document.cookie = arguments[0] + '=set';", key1);
            ((IJavaScriptExecutor)driver).ExecuteScript("document.cookie = arguments[0] + '=set';", key2);

            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed(key1));
            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed(key2));

            driver.Manage().Cookies.DeleteCookieNamed(key1);

            Assert.IsNull(driver.Manage().Cookies.GetCookieNamed(key1));
            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed(key2));
        }

        [Test]
        public void ShouldNotDeleteCookiesWithASimilarName()
        {
            string cookieOneName = "fish";
            Cookie cookie1 = new Cookie(cookieOneName, "cod");
            Cookie cookie2 = new Cookie(cookieOneName + "x", "earth");
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);
            options.Cookies.AddCookie(cookie2);

            options.Cookies.DeleteCookieNamed(cookieOneName);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;

            Assert.IsFalse(driver.Manage().Cookies.AllCookies.Contains(cookie1));
            Assert.IsTrue(driver.Manage().Cookies.AllCookies.Contains(cookie2));
        }

        [Test]
        public void AddCookiesWithDifferentPathsThatAreRelatedToOurs()
        {
            string basePath = EnvironmentManager.Instance.UrlBuilder.Path;

            Cookie cookie1 = new Cookie("fish", "cod", "/" + basePath + "/animals");
            Cookie cookie2 = new Cookie("planet", "earth", "/" + basePath + "/");
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);
            options.Cookies.AddCookie(cookie2);

            UrlBuilder builder = EnvironmentManager.Instance.UrlBuilder;
            driver.Url = builder.WhereIs("animals");

            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.IsTrue(cookies.Contains(cookie1));
            Assert.IsTrue(cookies.Contains(cookie2));

            driver.Url = builder.WhereIs("");
            cookies = options.Cookies.AllCookies;
            Assert.IsFalse(cookies.Contains(cookie1));
            Assert.IsTrue(cookies.Contains(cookie2));
        }

        [Test]
        public void CanSetCookiesOnADifferentPathOfTheSameHost()
        {
            string basePath = EnvironmentManager.Instance.UrlBuilder.Path;
            Cookie cookie1 = new Cookie("fish", "cod", "/" + basePath + "/animals");
            Cookie cookie2 = new Cookie("planet", "earth", "/" + basePath + "/galaxy");

            IOptions options = driver.Manage();
            ReadOnlyCollection<Cookie> count = options.Cookies.AllCookies;

            options.Cookies.AddCookie(cookie1);
            options.Cookies.AddCookie(cookie2);

            string url = EnvironmentManager.Instance.UrlBuilder.WhereIs("animals");
            driver.Url = url;
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;

            Assert.IsTrue(cookies.Contains(cookie1));
            Assert.IsFalse(cookies.Contains(cookie2));

            driver.Url = EnvironmentManager.Instance.UrlBuilder.WhereIs("galaxy");
            cookies = options.Cookies.AllCookies;
            Assert.IsFalse(cookies.Contains(cookie1));
            Assert.IsTrue(cookies.Contains(cookie2));
        }

        [Test]
        public void ShouldNotBeAbleToSetDomainToSomethingThatIsUnrelatedToTheCurrentDomain()
        {
            Cookie cookie1 = new Cookie("fish", "cod");
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);

            string url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("simpleTest.html");
            driver.Url = url;

            Assert.IsNull(options.Cookies.GetCookieNamed("fish"));
        }

        [Test]
        public void ShouldBeAbleToAddToADomainWhichIsRelatedToTheCurrentDomain()
        {
            Regex nameRegex = new Regex("\\d{1,3}(?:\\.\\d{1,3}){3}");
            string name = GotoValidDomainAndClearCookies();
            if (name == null || nameRegex.IsMatch(name))
            {
                Assert.Ignore("Skipping test: Cookies can only be set on fully-qualified domain names.");
            }

            Assert.IsNull(driver.Manage().Cookies.GetCookieNamed("name"));

            Regex replaceRegex = new Regex(".*?\\.");
            string shorter = replaceRegex.Replace(name, ".");
            Cookie cookie = new Cookie("name", "value", shorter, "/", DateTime.Now.AddSeconds(100000));

            driver.Manage().Cookies.AddCookie(cookie);

            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed("name"));
        }

        [Test]
        public void ShouldBeAbleToIncludeLeadingPeriodInDomainName()
        {
            Regex nameRegex = new Regex("\\d{1,3}(?:\\.\\d{1,3}){3}");
            string name = GotoValidDomainAndClearCookies();
            if (name == null || nameRegex.IsMatch(name))
            {
                Assert.Ignore("Skipping test: Cookies can only be set on fully-qualified domain names.");
            }
            driver.Manage().Cookies.DeleteAllCookies();

            Assert.IsNull(driver.Manage().Cookies.GetCookieNamed("name"), "Looks like delete all cookies doesn't");

            // Replace the first part of the name with a period
            Regex replaceRegex = new Regex(".*?\\.");
            string shorter = replaceRegex.Replace(name, ".");
            Cookie cookie = new Cookie("name", "value", shorter, "/", DateTime.Now.AddSeconds(100000));

            driver.Manage().Cookies.AddCookie(cookie);

            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed("name"));
        }

        [Test]
        public void GetCookieDoesNotRetriveBeyondCurrentDomain()
        {
            Cookie cookie1 = new Cookie("fish", "cod");
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);

            String url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("");
            driver.Url = url;

            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.IsFalse(cookies.Contains(cookie1));
        }

        [Test]
        [IgnoreBrowser(Browser.IE, "IE cookies do not conform to RFC, so setting cookie on domain fails.")]
        [IgnoreBrowser(Browser.Chrome, "Test tries to set the domain explicitly to 'localhost', which is not allowed.")]
        public void ShouldBeAbleToSetDomainToTheCurrentDomain()
        {
            Uri url = new Uri(driver.Url);
            String host = url.Host + ":" + url.Port.ToString();

            Cookie cookie1 = new Cookie("fish", "cod", host, "/", null);
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);

            driver.Url = javascriptPage;
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.IsTrue(cookies.Contains(cookie1));
        }

        [Test]
        public void ShouldWalkThePathToDeleteACookie()
        {
            string basePath = EnvironmentManager.Instance.UrlBuilder.Path;

            Cookie cookie1 = new Cookie("fish", "cod");
            driver.Manage().Cookies.AddCookie(cookie1);
            int count = driver.Manage().Cookies.AllCookies.Count;

            driver.Url = childPage;
            Cookie cookie2 = new Cookie("rodent", "hamster", "/" + basePath + "/child");
            driver.Manage().Cookies.AddCookie(cookie2);
            count = driver.Manage().Cookies.AllCookies.Count;

            driver.Url = grandchildPage;
            Cookie cookie3 = new Cookie("dog", "dalmation", "/" + basePath + "/child/grandchild/");
            driver.Manage().Cookies.AddCookie(cookie3);
            count = driver.Manage().Cookies.AllCookies.Count;

            driver.Url = (EnvironmentManager.Instance.UrlBuilder.WhereIs("child/grandchild"));
            driver.Manage().Cookies.DeleteCookieNamed("rodent");
            count = driver.Manage().Cookies.AllCookies.Count;

            Assert.IsNull(driver.Manage().Cookies.GetCookieNamed("rodent"));

            ReadOnlyCollection<Cookie> cookies = driver.Manage().Cookies.AllCookies;
            Assert.AreEqual(2, cookies.Count);
            Assert.IsTrue(cookies.Contains(cookie1));
            Assert.IsTrue(cookies.Contains(cookie3));

            driver.Manage().Cookies.DeleteAllCookies();
            driver.Url = grandchildPage;

            cookies = driver.Manage().Cookies.AllCookies;
            Assert.AreEqual(0, cookies.Count);
        }

        [Test]
        [IgnoreBrowser(Browser.IE, "IE cookies do not conform to RFC, so setting cookie on domain fails.")]
        [IgnoreBrowser(Browser.Chrome, "Test tries to set the domain explicitly to 'localhost', which is not allowed.")]
        public void ShouldIgnoreThePortNumberOfTheHostWhenSettingTheCookie()
        {
            Uri uri = new Uri(driver.Url);
            String host = string.Format("{0}:{1}", uri.Host, uri.Port);

            Assert.IsNull(driver.Manage().Cookies.GetCookieNamed("name"));
            Cookie cookie = new Cookie("name", "value", host, "/", null);
            driver.Manage().Cookies.AddCookie(cookie);

            Assert.IsNotNull(driver.Manage().Cookies.GetCookieNamed("name"));
        }

        [Test]
        public void CookieIntegrity()
        {
            string url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("animals");

            driver.Url = url;
            driver.Manage().Cookies.DeleteAllCookies();

            DateTime time = DateTime.Now.AddDays(1);
            Cookie cookie1 = new Cookie("fish", "cod", null, "/common/animals", time);
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie1);

            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Cookie retrievedCookie = null;
            foreach (Cookie tempCookie in cookies)
            {
                if (cookie1.Equals(tempCookie))
                {
                    retrievedCookie = tempCookie;
                    break;
                }
            }

            Assert.IsNotNull(retrievedCookie);
            //Cookie.equals only compares name, domain and path
            Assert.AreEqual(cookie1, retrievedCookie);
        }

        [Test]
        [IgnoreBrowser(Browser.Chrome, "Chrome and Selenium, which use JavaScript to retrieve cookies, cannot return expiry info;")]
        [IgnoreBrowser(Browser.IE, "IE does not return expiry info")]
        [IgnoreBrowser(Browser.Android, "Chrome and Selenium, which use JavaScript to retrieve cookies, cannot return expiry info;")]
        public void ShouldRetainCookieExpiry()
        {
            string url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("animals");

            driver.Url = url;
            driver.Manage().Cookies.DeleteAllCookies();

            // DateTime.Now contains milliseconds; the returned cookie expire date
            // will not. So we need to truncate the milliseconds.
            DateTime current = DateTime.Now;
            DateTime expireDate = new DateTime(current.Year, current.Month, current.Day, current.Hour, current.Minute, current.Second, DateTimeKind.Local).AddDays(1);

            Cookie addCookie = new Cookie("fish", "cod", "/common/animals", expireDate);
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(addCookie);

            Cookie retrieved = options.Cookies.GetCookieNamed("fish");
            Assert.IsNotNull(retrieved);
            Assert.AreEqual(addCookie.Expiry, retrieved.Expiry, "Cookies are not equal");
        }

        [Test]
        public void SettingACookieThatExpiredInThePast()
        {
            string url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("animals");

            driver.Url = url;
            driver.Manage().Cookies.DeleteAllCookies();

            DateTime expires = DateTime.Now.AddSeconds(-1000);
            Cookie cookie = new Cookie("expired", "yes", "/common/animals", expires);
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(cookie);

            cookie = options.Cookies.GetCookieNamed("fish");
            Assert.IsNull(cookie, "Cookie expired before it was set, so nothing should be returned: " + cookie);
        }

        //////////////////////////////////////////////
        // Tests unique to the .NET language bindings
        //////////////////////////////////////////////

        [Test]
        public void ShouldAddCookieToCurrentDomainAndPath()
        {
            // Cookies cannot be set on domain names with less than 2 dots, so
            // localhost is out. If we are in that boat, bail the test.
            string hostName = EnvironmentManager.Instance.UrlBuilder.HostName;
            string[] hostNameParts = hostName.Split(new char[] { '.' });
            if (hostNameParts.Length < 3)
            {
                Assert.Ignore("Skipping test: Cookies can only be set on fully-qualified domain names.");
            }
            driver.Url = macbethPage;
            IOptions options = driver.Manage();
            Cookie cookie = new Cookie("Homer", "Simpson", hostName, "/" + EnvironmentManager.Instance.UrlBuilder.Path, null);
            options.Cookies.AddCookie(cookie);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.That(cookies.Contains(cookie), "Valid cookie was not returned");
        }

        [Test]
        [IgnoreBrowser(Browser.IE, "Add cookie to unrelated domain silently fails for IE.")]
        [ExpectedException(typeof(WebDriverException))]
        public void ShouldNotShowCookieAddedToDifferentDomain()
        {
            driver.Url = macbethPage;
            IOptions options = driver.Manage();
            Cookie cookie = new Cookie("Bart", "Simpson", EnvironmentManager.Instance.UrlBuilder.HostName + ".com", EnvironmentManager.Instance.UrlBuilder.Path, null);
            options.Cookies.AddCookie(cookie);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.IsFalse(cookies.Contains(cookie), "Invalid cookie was returned");
        }

        [Test]
        [IgnoreBrowser(Browser.Chrome, "Test tries to set the domain explicitly to 'localhost', which is not allowed.")]
        public void ShouldNotShowCookieAddedToDifferentPath()
        {
            driver.Url = macbethPage;
            IOptions options = driver.Manage();
            Cookie cookie = new Cookie("Lisa", "Simpson", EnvironmentManager.Instance.UrlBuilder.HostName, "/" + EnvironmentManager.Instance.UrlBuilder.Path + "IDoNotExist", null);
            options.Cookies.AddCookie(cookie);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.IsFalse(cookies.Contains(cookie), "Invalid cookie was returned");
        }

        // TODO(JimEvans): Disabling this test for now. If your network is using
        // something like OpenDNS or Google DNS which you may be automatically
        // redirected to a search page, which will be a valid page and will allow a
        // cookie to be created. Need to investigate further.
        // [Test]
        // [ExpectedException(typeof(InvalidOperationException))]
        public void ShouldThrowExceptionWhenAddingCookieToNonExistingDomain()
        {
            driver.Url = macbethPage;
            driver.Url = "doesnot.noireallyreallyreallydontexist.com";
            IOptions options = driver.Manage();
            Cookie cookie = new Cookie("question", "dunno");
            options.Cookies.AddCookie(cookie);
        }

        [Test]
        [IgnoreBrowser(Browser.Chrome, "Chrome and Selenium, which use JavaScript to retrieve cookies, cannot return expiry info;")]
        public void ShouldReturnNullBecauseCookieRetainsExpiry()
        {
            string url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("animals");
            driver.Url = url;

            driver.Manage().Cookies.DeleteAllCookies();

            Cookie addCookie = new Cookie("fish", "cod", "/common/animals", DateTime.Now.AddHours(-1));
            IOptions options = driver.Manage();
            options.Cookies.AddCookie(addCookie);

            Cookie retrieved = options.Cookies.GetCookieNamed("fish");
            Assert.IsNull(retrieved);
        }

        [Test]
        public void ShouldAddCookieToCurrentDomain()
        {
            driver.Url = macbethPage;
            IOptions options = driver.Manage();
            Cookie cookie = new Cookie("Marge", "Simpson", "/");
            options.Cookies.AddCookie(cookie);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            Assert.That(cookies.Contains(cookie), "Valid cookie was not returned");
        }

        [Test]
        public void ShouldDeleteCookie()
        {
            driver.Url = macbethPage;
            IOptions options = driver.Manage();
            Cookie cookieToDelete = new Cookie("answer", "42");
            Cookie cookieToKeep = new Cookie("canIHaz", "Cheeseburguer");
            options.Cookies.AddCookie(cookieToDelete);
            options.Cookies.AddCookie(cookieToKeep);
            ReadOnlyCollection<Cookie> cookies = options.Cookies.AllCookies;
            options.Cookies.DeleteCookie(cookieToDelete);
            ReadOnlyCollection<Cookie> cookies2 = options.Cookies.AllCookies;
            Assert.IsFalse(cookies2.Contains(cookieToDelete), "Cookie was not deleted successfully");
            Assert.That(cookies2.Contains(cookieToKeep), "Valid cookie was not returned");
        }

        private string GotoValidDomainAndClearCookies()
        {
            Regex hostNameRegex = new Regex("\\w+\\.\\w+.*");

            string name = null;
            string hostName = EnvironmentManager.Instance.UrlBuilder.HostName;
            if (hostNameRegex.IsMatch(hostName))
            {
                name = hostName;
                driver.Url = EnvironmentManager.Instance.UrlBuilder.WhereIs("simpleTest.html");
            }
            hostName = EnvironmentManager.Instance.UrlBuilder.AlternateHostName;
            if (name != null && hostNameRegex.IsMatch(hostName))
            {
                name = hostName;
                driver.Url = EnvironmentManager.Instance.UrlBuilder.WhereElseIs("simpleTest.html");
            }

            driver.Manage().Cookies.DeleteAllCookies();

            return name;
        }
    }
}
